File: 算法/3. 无重复字符的最长子串/LengthOfLongestSubstring.go
package ___无重复字符的最长子串

func lengthOfLongestSubstring(s string) int {
	left, right, maxValue := 0, 0, 0
	memo := make(map[uint8]int)
	for right < len(s) {
		value := s[right]
		memo[value] = memo[value] + 1
		for memo[value] > 1 {
			leftValue := s[left]
			memo[leftValue] = memo[leftValue] - 1
			left++
		}
		maxValue = max(maxValue, right-left+1)
		right++
	}
	return maxValue
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: 算法/3. 无重复字符的最长子串/LengthOfLongestSubstring.java
import java.util.HashMap;
import java.util.Map;

public class LengthOfLongestSubstring {

    class Solution {
        public int lengthOfLongestSubstring(String s) {
            if (s == null || s.length() == 0) {
                return 0;
            }
            Map<Character, Integer> memo = new HashMap<>();
            int result = 0;
            int left = 0;
            int right = 0;
            // 循环查询最长不重复字符串
            for (int i = 0; i < s.length(); i++) {
                char value = s.charAt(i);
                memo.put(value, memo.getOrDefault(value, 0) + 1);
                // 判断是否有重复，有的话left到不重复为止
                while (memo.get(value) > 1) {
                    char leftValue = s.charAt(left);
                    // 计数扣减
                    memo.put(leftValue, memo.getOrDefault(leftValue, 0) - 1);
                    left++;
                }
                result = Math.max(result, right - left + 1);
                right++;
            }
            return result;
        }
    }

}
